### ISSUE TITLE
[CI] NodeShutdownIT testShutdownReadinessService failing

### ISSUE THREAD

**Issue by luigidellaquila** [elastic/elasticsearch#91682]
**Build scan:**
https://gradle-enterprise.elastic.co/s/wjstsypdoyez6/tests/:x-pack:plugin:shutdown:qa:multi-node:javaRestTest/org.elasticsearch.xpack.shutdown.NodeShutdownIT/testShutdownReadinessService

**Reproduction line:**
```
./gradlew ':x-pack:plugin:shutdown:qa:multi-node:javaRestTest' --tests "org.elasticsearch.xpack.shutdown.NodeShutdownIT.testShutdownReadinessService" -Dtests.seed=E1262674CB60A393 -Dtests.locale=ar-MA -Dtests.timezone=America/Kralendijk -Druntime.java=17 -Dtests.fips.enabled=true
```

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.shutdown.NodeShutdownIT&tests.test=testShutdownReadinessService

**Failure excerpt:**
```
junit.framework.AssertionFailedError: Expected exception IOException but no exception was thrown

  at __randomizedtesting.SeedInfo.seed([E1262674CB60A393:2C6B3F0E690FF71B]:0)
  at org.apache.lucene.tests.util.LuceneTestCase.expectThrows(LuceneTestCase.java:2854)
  at org.apache.lucene.tests.util.LuceneTestCase.expectThrows(LuceneTestCase.java:2840)
  at org.elasticsearch.test.readiness.ReadinessClientProbe.lambda$tcpReadinessProbeFalse$2(ReadinessClientProbe.java:73)
  at java.security.AccessController.doPrivileged(AccessController.java:318)
  at org.elasticsearch.test.readiness.ReadinessClientProbe.tcpReadinessProbeFalse(ReadinessClientProbe.java:72)
  at org.elasticsearch.xpack.shutdown.NodeShutdownIT.testShutdownReadinessService(NodeShutdownIT.java:133)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:833)

```

**Comment by thecoop** [comment]
This is the same exception as #91434

**Comment by pgomulka** [comment]
the exception is slightly different. In my view the problem is the same as 91434 but affect a different test. I will reopen so that we don't forget fix this test as well
```
  at org.elasticsearch.xpack.shutdown.NodeShutdownIT.testShutdownReadinessService(NodeShutdownIT.java:133)
```

**Comment by ldematte** [comment]
I spent some time understanding this one better. 
The test is a rest test, so we do not have direct access to the cluster - the solution in https://github.com/elastic/elasticsearch/pull/98994 does not apply directly.
The test test plugin [set up a system variable containing the test ports](https://github.com/elastic/elasticsearch/blob/d1f2b5d79ddfae4e63777960f1406931abca80c5/build-tools/src/main/java/org/elasticsearch/gradle/test/JavaRestTestPlugin.java#L65); however, the port is still ephemeral.

I can see two ways in which this test could fail:
1. Like for https://github.com/elastic/elasticsearch/issues/91434, ephemeral ports may be an issue (see that issue for details)
2. There might be a delay/race condition. The request for cluster shutdown and probe for readiness are back-to-back; however the readiness service stops (synchronously - sync with the listener thread appears correct) as a [response to a change in cluster state](https://github.com/elastic/elasticsearch/blob/f79257687d5d32cecd85a6e656915c8b775e1c01/server/src/main/java/org/elasticsearch/readiness/ReadinessService.java#L237), which happens on a different executor (in [ClusterApplierService](https://github.com/elastic/elasticsearch/blob/a20ee3f8f268a4f181b3e898635a96a3f605721a/server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java#L306))

Both are unlikely but can happen. The first one is a test framework issue - to solve it I'd consider moving this test to an internalClusterTest and use the same mocks introduced in https://github.com/elastic/elasticsearch/pull/98994

The second one may or may not be an issue - @pgomulka @thecoop do we care that the cluster readiness is immediately changed?
If not, a not elegant but effective solution may be to repeat the test for a fixed number of retries with some delay between them, and fail or proceed with the test only after.

**Comment by ldematte** [comment]
I propose to move the test to an internalClusterTest anyway.
If the asynchronicity between Shutdown and Readiness is not an issue, introduce the delayed retries too. Otherwise we should open a proper bug for that.

**Comment by pgomulka** [comment]
hey @ldematte last time when I was looking into this I suspected the 1.
I was thinking that both these test coudl be running on the same machine so the readiness check expecting false in this test (now line 140) was returning true because of readiness service running in ReadinessClusterIT

**Comment by ldematte** [comment]
Yes, @pgomulka, I agree this is the most likely. But 2. can happen too, and I would like to know if it is OK or not.
In any case, 2. is even more unlikely to happen, while 1. as we have seen in other tests have a small but not trivial chance to happen. 
So I propose we move this test to an internalClusterTest and use the same mocks introduced in https://github.com/elastic/elasticsearch/pull/98994 anyway. 
If 2. is an issue, I think is better to open a separate issue. If it's not, we can introduce a small and simple retry-with-delay in the test.

**Comment by pgomulka** [comment]
agree with your plan
at the same time I think the likelihood of 1 has decreased now as the ReadinessClusterIT is no longer opening ports, right?

**Comment by ldematte** [comment]
Yes; on the other hand, since the readiness check is just a connect on a given port, if the port is ephemeral any service binding on that port will make the cluster seem ready, even if it's another service entirely. 
(Maybe readiness should at least exchange a magic byte?)
